### PR TITLE
Added -help option to sourcekitd-test

### PR DESCRIPTION
**Pull request by felix91gr**
### What the PR includes
Addresses one of the current starter bugs, adding the feature mentioned in the title.

I've added a `FIXME:` in the **printHelp** function to suggest a possible expansion for its functionality (allowing to hide/include certain options from the Help Text). I don't think that's necessary for this command, but could be useful if we want it to have as many options as `Driver.cpp` has. 

### Where has this been tested

I've compiled this in Linux, using a slightly modified version of the default presets. I used #8485 and #8189, and added a `cmake` option to compile `sourcekitd-test`. It passed the **buildbot_linux, smoketest** tests on my pc :+1: 

Resolves [SR-1421](https://bugs.swift.org/browse/SR-1421).

### CODE REVIEW

**Comment by felix91gr** [comment]
@swift-ci please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by nkcsgexi** [comment]
Can we add a test to ensure this dump the -h option itself?

**Comment by felix91gr** [comment]
🤔 sure... I can add that

**Comment by felix91gr** [comment]
Do we have a documentation on the testing tools? I'm talking about the kind of tests found in swift/test

**Comment by felix91gr** [comment]
@nkcsgexi I made this test and it passed. Should I test something else as well?
```swift

// Basic check that usage works
//
// RUN: %sourcekitd-test -help | %FileCheck %s
//
// CHECK: USAGE: sourcekitd-test [options] <inputs>
```

**Comment by nkcsgexi** [comment]
Thank you! @swift-ci please smoke test

**Review by nkcsgexi — APPROVED** [review]

**Comment by felix91gr** [comment]
Thank you too🙂

Btw... 🤔 I made a new test, for checking that entering an unknown argument
prints an error message and suggests using -help. But running the 9xxx
tests takes a long time 😆.

Is there a way to run just *one* test? Or to run just the SourceKit ones?

**Comment by nkcsgexi** [comment]
@felix91gr could you push your test and ask swift CI to test it for you?

**Comment by felix91gr** [comment]
Sure! I'll push it 🙂

I can't trigger swift CI though... I'm a new-ish member of the community,
so I haven't been given access to it.

(I already tried in my first comment, that was when I learned)

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by felix91gr** [comment]
> please smoke test

Thanks! ^^

**Comment by nkcsgexi** [comment]
@felix91gr whenever you need to trigger CI, just comment on the pull request to ask some one to trigger it for you. I'm sure many are glad to help 🙂

**Comment by felix91gr** [comment]
It failed, this is the output:
```console
Command Output (stderr):
--
error: unknown argument: -this_option_does_not_exist
Use -h or -help for assistance
FileCheck error: '-' is empty.
FileCheck command line:  /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/llvm-macosx-x86_64/./bin/FileCheck /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/SourceKit/Misc/wrong_arguments.swift

--
```
From that, only the first two lines are the output of `sourcekitd-test`. That's the correct output (what we are testing, in fact). I have problems interpreting the `FileCheck` error though... what does...

 `FileCheck error: '-' is empty.` mean?

This is the original test, btw:

```swift
// RUN: %sourcekitd-test -this_option_does_not_exist | %FileCheck %s
//
// CHECK: error: unknown argument:
// CHECK: Use -h or -help for assistance
```

Should I use the complete line? (like `// CHECK: error: unknown argument: -this_option_does_not_exist`)

I thought `CHECK` didn't enforce the line to be exactly equal... I wanted to give freedom to have any erroneous option in the test, just in case. 

### (Meta)

:thinking: it failed only on OS X... Now that I look at it, the entire SourceKit test suite is disabled in Linux.

Ah, _right_. We haven't yet merged Norio Nomura's PR #8485. Well, I'll test it on my machine as well then (I'm using Linux) :+1:

**Comment by felix91gr** [comment]
Yup, the `CHECK` string must be matched exactly. That must've been the error.

@nkcsgexi I've submitted a correction. Can you call Swift-CI please? :)

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by felix91gr** [comment]
:thinking: both failed... that's weird.

Well, I'll restart my setup and keep at it.

**Comment by felix91gr** [comment]
@benlangmuir or @nkcsgexi, I think I fixed the tests this time... sorry to bother you, could you run a smoke test please? :)

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by felix91gr** [comment]
@benlangmuir or @nkcsgexi... Sorry to bother you again. This time I've tested the changes using a local version of `FileCheck`. It _should_ work. If it doesn't... :thinking: we'll see.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by felix91gr** [comment]
Wohoo! The `wrong_arguments.swift` test passed! I'm making it work :D 

Next up, I'll fix whatever is wrong with the `usage.swift` test :fist_raised:

**Comment by felix91gr** [comment]
I think I know what's the problem. I'll upload the change.

**Comment by felix91gr** [comment]
The problem was that I thought the `-help` command returned 0, because it doesn't trigger an error. **I was wrong**. 

In the design of `sourcekitd-test`, **any option** that makes the program not test a SourceKit request, **returns 1**. If the **"happy path"** would be to parse a request and perform a SourceKit test on it, then **anything but the happy path returns 1**. Therefore, both of my tests (`wrong_arguments` and `usage` should expect a 1 for result, and therefore should start with `{// RUN: not sourcekitd-test ...}`. Which now, they do :) _(I tried it first with `wrong_arguments`, and it passes)_

So, for (hopefully) last time... @nkcsgexi could you call CI? :)

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by felix91gr** [comment]
Aw yeah! It passed! :D

**Comment by felix91gr** [comment]
It passed! 
Thanks for your patience, @nkcsgexi :smile: 

It was a dumb mistake, as it always is :D

**Comment by felix91gr** [comment]
Should we merge this? Should I add something else? 🙂

**Comment by nkcsgexi** [comment]
Could you squash all commits into one?

**Comment by felix91gr** [comment]
Oh, good idea. Will do.

**Comment by felix91gr** [comment]
Done! :)

**Comment by nkcsgexi** [comment]
👍 @swift-ci please smoke test and merge

**Comment by felix91gr** [comment]
🤔 seems like nothing happened...

**Comment by nkcsgexi** [comment]
Let's try again. @swift-ci please smoke test

**Comment by felix91gr** [comment]
Ahhh... you know, probably when asked for a test + merge process, it puts the task way back into the queue because the tasks before it where called using the previous state of the repo :D 

If that's the case, then it should take a while sometimes, depending on how busy CI is :thinking:

File: utils/build-presets.ini [hunk 1]
```diff
@@ -708,7 +708,7 @@ install-llbuild
 install-swiftpm
 install-xctest
 install-prefix=/usr
-swift-install-components=autolink-driver;compiler;clang-builtin-headers;stdlib;swift-remote-mirror;sdk-overlay;license
+swift-install-components=autolink-driver;compiler;clang-builtin-headers;stdlib;swift-remote-mirror;sdk-overlay;license;sourcekit-inproc
```
Comment (akyrtzi) [inline utils/build-presets.ini]: This seems irrelevant, is this change intentional ?

Comment (felix91gr) [inline utils/build-presets.ini]: Oh. I didn't notice. It's a leftover of one of my cherrypicks from Nomura's PR (#8485). Helped me test things in linux.
It can, and probably should, be removed... even though that PR has already been aprooved. You can do it if you want, or just wait for me (atm I'm not at the computer) :)

Comment (akyrtzi) [inline utils/build-presets.ini]: Remove the change when you can, and afterwards we can merge your improvements.

Comment (felix91gr) [inline utils/build-presets.ini]: Oki doke 🙂👍🏻

**Comment by felix91gr** [comment]
@akyrtzi done :wink:

**Comment by felix91gr** [comment]
#### Updated
* I deleted everything that was not needed:
  * A `//WIP` comment that I made in `TestOptions.cpp`, and 
  * The `;sourcekit-inproc` install option in the build presets that I used for testing. 
* I also added:
  * A brief documentation on the new tests, and 
  * A _newline_ that was missing at the end of `TestOptions.cpp`

I think we're ready for a test and merge. @nkcsgexi or @akyrtzi could you please call CI again? :)

**Comment by akyrtzi** [comment]
@swift-ci smoke test and merge

**Comment by akyrtzi** [comment]
@swift-ci smoke test

**Comment by felix91gr** [comment]
What a weird thing... why did it fail the [smoke test & merge](https://ci.swift.org/view/Pull%20Request/job/swift-PR-smoke-merge/2615/console), but passed the smoke test alone?

The only thing I can think of is that I haven't set gpg keys for my Github account. That could be it :thinking:

**Comment by felix91gr** [comment]
No, wait... it says "wrong SHA". :thinking:
What made the SHA change?

**Comment by felix91gr** [comment]
:smile: 

Thanks for the help and the patience, guys! @nkcsgexi and @akyrtzi specially :)

**Comment by nkcsgexi** [comment]
@felix91gr congrats for your first Swift contribution! 👍
